server/src/main/java/org/opensearch/search/SearchService.java

```diff
@@ -1114,6 +1132,22 @@ private DefaultSearchContext createSearchContext(ReaderContext reader, ShardSear
         return searchContext;
     }
 
+    private void writeTaskResourceUsage(SearchShardTask task) {
+        try {
+            // Get current resource usages from when the task started
+            TaskResourceInfo.Builder builder = taskResourceTrackingService.getCurrentTaskResourceUsageBuilder(task);
+            if (builder != null) {
+                // Attach NodeId to taskResourceInfo
+                TaskResourceInfo taskResourceInfo = builder.setNodeId(clusterService.localNode().getId()).build();
+                // Remove the existing TASK_RESOURCE_USAGE header since it would have come from an earlier phase in the same request.
+                threadPool.getThreadContext().removeResponseHeader(TASK_RESOURCE_USAGE);
+                threadPool.getThreadContext().addResponseHeader(TASK_RESOURCE_USAGE, taskResourceInfo.toString());
```

Comment:
SearchService class should be completely agnostic of the headers being massaged for tracking information. Can we further refactor this logic? Maybe even move the whole writeTaskResourceUsage method in trackingService?

Reply:
just updated!